Reduce allocations in RuntimeGraph

drewnoakes opened the pull request:
## Bug

Fixes: https://github.com/NuGet/Home/issues/12717
Fixes: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1826702
Fixes: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1826712

Regression? Last working version: N/A

## Description

- `RuntimeGraph` objects are immutable. The `Clone` operation can just return itself for such values. Similarly, the `Merge` operation can be optimised in cases where either operand is empty.
- Don't initialize the three `ConcurrentDictionary<>` fields on `RuntimeGraph` when empty.
- Share the `ReadOnlyDictionary<>` instances for `Runtimes` and `Supports` properties across all empty instances by ensuring there's only a single instance that's empty, via the `Empty` property and making the parameterless constructor private.
- Remove the `Supports` property setter as this type should be immutable. The property appeared unused.
- Null annotate the class and update the API files accordingly.
- Short circuit several methods when `IsEmpty` is `true`.
- Add code to ensure that `IsEmpty` is checked before trying to dereference cache fields, that are only initialised when not empty. To help constrain these call sites in a more obvious way, the `GetOrAdd`-style callbacks were moved to local functions.
- Convert dictionary key classes to `readonly struct` types to reduce heap allocations.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  <!-- Describe why you haven't added automation. -->
  - [ ] Test exception
  - **OR**
  - [x] N/A <!-- Infrastructure, documentation etc. -->

- **Documentation**
  <!-- Please link the PR/issue if appropriate -->
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

drewnoakes committed:
RuntimeGraph.Clone/Merge don't allocate when empty

`RuntimeGraph` objects are immutable. The `Clone` operation can just return itself for such values. Similarly, the `Merge` operation can be optimised in cases where either operand is empty.

davkean commented on src/NuGet.Core/NuGet.Packaging/RuntimeModel/RuntimeGraph.cs:
```diff
@@ -10,7 +10,7 @@
 namespace NuGet.RuntimeModel
 {
-    public class RuntimeGraph : IEquatable<RuntimeGraph>
+    public sealed class RuntimeGraph : IEquatable<RuntimeGraph>
```
Note, this is a breaking change - but given how NuGet objects implement Clone where they assume no derived classes, they should probably be sealed by default.

drewnoakes committed:
Make an empty RuntimeGraph more lightweight

Don't initialize the three `ConcurrentDictionary<>` fields on `RuntimeGraph` when empty. Also, share the `ReadOnlyDictionary<>` instances for `Runtimes` and `Supports` properties across all empty instances by ensuring there's only a single instance that's empty, via the `Empty` property and making the parameterless constructor private.

Also, removed the `Supports` property setter as this type should be immutable. The property appeared unused.

Annotate the class and update the API files accordingly.

Add code to ensure that `IsEmpty` is checked before trying to dereference cache fields, that are only initialised for non-empty objects. Many of those calls can be short-circuited for empty objects. To help constrain these call sites in a more obvious way, the `GetOrAdd`-style callbacks were moved to local functions.

drewnoakes commented:
Found some more wins, and a fix for https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1826702. I pushed an additional commit that invalidates @davkean's review. PR description updated accordingly.

drewnoakes commented on src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PackageBuilder.cs:
```diff
@@ -809,7 +809,7 @@ private static void ValidateFileFrameworks(IEnumerable<IPackageFile> files)
             }
 
-            var managedCodeConventions = new ManagedCodeConventions(new RuntimeGraph());
+            var managedCodeConventions = new ManagedCodeConventions(RuntimeGraph.Empty);
```
We now have a singleton empty instance.

drewnoakes commented on src/NuGet.Core/NuGet.Packaging/RuntimeModel/RuntimeGraph.cs:
```diff
@@ -8,31 +8,33 @@
 using System.Linq;
 using NuGet.Shared;
 
+#nullable enable
```
This file has been annotated, resulting in a large diff in the API text files.

drewnoakes commented on src/NuGet.Core/NuGet.Packaging/RuntimeModel/RuntimeGraph.cs:
```diff
@@ -8,31 +8,33 @@
+        public static readonly RuntimeGraph Empty = new();
 
         public IReadOnlyDictionary<string, RuntimeDescription> Runtimes { get; }
-        public IReadOnlyDictionary<string, CompatibilityProfile> Supports { get; set; }
+        public IReadOnlyDictionary<string, CompatibilityProfile> Supports { get; }
```
This setter was removed. Note that `Runtimes` didn't have one to remove. These objects are otherwise immutable, and the optimisations here require that immutability. There were no usages of this setter in this repo.

nkolev92 replied:
Unclear to me why https://github.com/NuGet/NuGet.Client/commit/3dc89fbb2b29ca8f05b797d8d49440931e45b348 added a setter. It wasn't used in that PR either.

drewnoakes commented on src/NuGet.Core/NuGet.Packaging/RuntimeModel/RuntimeGraph.cs:
```diff
@@ -54,10 +56,21 @@ private RuntimeGraph(Dictionary<string, RuntimeDescription> runtimes, Dictionary
+
         public RuntimeGraph Clone()
         {
+            if (IsEmpty)
+            {
+                return this;
+            }
```
Optimise clone for empty.

drewnoakes commented on src/NuGet.Core/NuGet.Packaging/RuntimeModel/RuntimeGraph.cs:
```diff
@@ -67,6 +80,16 @@ public RuntimeGraph Clone()
         /// <param name="other">The other graph to merge in to this graph</param>
         public static RuntimeGraph Merge(RuntimeGraph left, RuntimeGraph right)
         {
+            if (left.IsEmpty)
+            {
+                return right;
+            }
+
+            if (right.IsEmpty)
+            {
+                return left;
+            }
```
Optimise merge when either is empty.

drewnoakes commented on src/NuGet.Core/NuGet.Packaging/RuntimeModel/RuntimeGraph.cs:
```diff
@@ -110,50 +132,52 @@ public static RuntimeGraph Merge(RuntimeGraph left, RuntimeGraph right)
-        /// Expand runtimes in a BFS walk. This ensures that nearest RIDs are returned first.
-        /// Ordering is important for finding the nearest runtime dependency.
-        /// </summary>
-        private IEnumerable<string> ExpandRuntimeInternal(string runtime)
```
This method was converted to a local function, as it should only be called on non-empty instances. All callers of `ExpandRuntimeCached` validate the instance is non-empty before calling. The method is otherwise unchanged.

drewnoakes commented on src/NuGet.Core/NuGet.Packaging/RuntimeModel/RuntimeGraph.cs:
```diff
@@ -173,18 +197,28 @@ public bool AreCompatible(string criteria, string provided)
-        }
+            return _areCompatible!.GetOrAdd(key, AreCompatibleInternal);
 
-        private bool AreCompatibleInternal(RuntimeCompatKey key)
```
Ditto, local function.

drewnoakes commented on src/NuGet.Core/NuGet.Packaging/RuntimeModel/RuntimeGraph.cs:
```diff
@@ -173,18 +197,28 @@ public bool AreCompatible(string criteria, string provided)
 
         public IEnumerable<RuntimePackageDependency> FindRuntimeDependencies(string runtimeName, string packageId)
         {
+            if (IsEmpty)
+            {
+                return Enumerable.Empty<RuntimePackageDependency>();
+            }
```
Optimise when empty.

drewnoakes commented on src/NuGet.Core/NuGet.Packaging/RuntimeModel/RuntimeGraph.cs:
```diff
@@ -173,18 +197,28 @@ public bool AreCompatible(string criteria, string provided)
                 return true;
             }
 
+            if (IsEmpty)
+            {
+                return false;
+            }
```
Optimise when empty.

drewnoakes commented on src/NuGet.Core/NuGet.Packaging/RuntimeModel/RuntimeGraph.cs:
```diff
@@ -309,7 +339,7 @@ public static void ReleaseList(List<T> list)
         /// <summary>
         /// RID + package id
         /// </summary>
-        private sealed class RuntimeDependencyKey : IEquatable<RuntimeDependencyKey>
+        private readonly struct RuntimeDependencyKey : IEquatable<RuntimeDependencyKey>
```
Convert this private collection key to a struct to reduce heap allocations.

drewnoakes commented on src/NuGet.Core/NuGet.Packaging/RuntimeModel/RuntimeGraph.cs:
```diff
@@ -341,7 +371,7 @@ public override int GetHashCode()
         /// <summary>
         /// RID -> RID compatibility key
         /// </summary>
-        private sealed class RuntimeCompatKey : IEquatable<RuntimeCompatKey>
+        private readonly struct RuntimeCompatKey : IEquatable<RuntimeCompatKey>
```
Convert this private collection key to a struct to reduce heap allocations.

kartheekp-ms replied:
Today, I learned about the concept of a readonly struct. According to the documentation, a `readonly struct` can undergo [boxing and unboxing conversions](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/struct#conversions) when it implements an interface. In this case, the `RuntimeCompatKey` is used as a key in a dictionary, and it implements the `IEquatable` interface. Therefore, I would anticipate that there might be some boxing operations involved when comparing keys for equality. Is it acceptable for these allocations to occur due to boxing?

drewnoakes replied:
Structs are only boxed when they're called through the interface.

```c#
int i = 1;
i.CompareTo(2); // call (non boxing and cheap)
IComparable j = i; // box
j.CompareTo(2); // callvirt and boxed argument (expensive)
```

Here's the MSIL for the above code.

```
// i.CompareTo(2)
        IL_0002: ldloca.s 0
        IL_0004: ldc.i4.2
        IL_0005: call instance int32 [mscorlib]System.Int32::CompareTo(int32)
        IL_000a: pop
// IComparable j = i
        IL_000b: ldloc.0
        IL_000c: box [mscorlib]System.Int32
// j.CompareTo(2)
        IL_0011: ldc.i4.2
        IL_0012: box [mscorlib]System.Int32
        IL_0017: callvirt instance int32 [mscorlib]System.IComparable::CompareTo(object)
        IL_001c: pop
        IL_001d: ret
```

In the case of the dictionary, if `TKey` is a struct, the JIT generates non-boxing calls to the `Equals` and `GetHashCode` overridden from `object`.

We could probably remove the `IEquatable<>` interfaces from this type and there'd be no difference in behaviour.

drewnoakes reviewed (commented):
Highlighting some of the diff here to help reviewers.

nkolev92 commented on src/NuGet.Core/NuGet.Packaging/RuntimeModel/RuntimeGraph.cs:
```diff
@@ -8,31 +8,33 @@
+        /// <summary>
+        /// Gets a singleton, immutable, empty instance of <see cref="RuntimeGraph"/>.
+        /// </summary>
+        public static readonly RuntimeGraph Empty = new();
```
nit: personally I like not modifying lines for minor stylistic changes. It just makes the diff unnecessarily larger.

nkolev92 reviewed (dismissed):
LGTM. 

Worth waiting for @zivkan and @jeffkl who are familiar with this code to review as well. 

I don't mind the setter breaking change, but maybe we just revert the sealed change?

drewnoakes commented:
> maybe we just revert the sealed change

The rationale here is that for any derived class the clone method cannot work correctly. I don't feel strongly either way and can remove `sealed` if preferred.

nkolev92 commented:
> > maybe we just revert the sealed change
> 
> The rationale here is that for any derived class the clone method cannot work correctly. I don't feel strongly either way and can remove `sealed` if preferred.

Don't feel too strongly. 
I don't expect this to cause any problems. 

dotnet/runtime is the only known user of these types I know, https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.NETCore.Platforms/src/GenerateRuntimeGraph.cs.

drewnoakes commented:
Ah ok, then this PR will break that code. I removed the parameterless `RuntimeGraph` constructor used here:

https://github.com/dotnet/runtime/blob/64cab8fd72b72f5ce708e5866fa4323ddf2613b4/src/libraries/Microsoft.NETCore.Platforms/src/GenerateRuntimeGraph.cs#L152

It's an easy fix (`new RuntimeGraph()` -> `RuntimeGraph.Empty`) but it may still cause binary issues. I'll reinstate it as public.

drewnoakes committed:
Reinstate RuntimeGraph constructor as public

nkolev92 commented:
I think we should try not to break them. I know bringing the constructor back is gonna come with some drawbacks, but I think that's where we're at right now. 

You can consider marking the constructor as obsolete.

drewnoakes committed:
RuntimeGraph constructors allocate less

drewnoakes commented:
I pushed an update that reinstates the constructor and also reduces allocations across the various constructors by reusing empty immutable `ReadOnlyDictionary<>` instances where possible.

> I know bringing the constructor back is gonna come with some drawbacks

With the update the drawback is minimal enough that I didn't mark it as obsolete.

nkolev92 commented:
There's some compile errors.

drewnoakes committed:
RuntimeGraph.Clone allocates less

drewnoakes committed:
Document RuntimeGraph properties

drewnoakes committed:
RuntimeGraph.Merge allocates less

drewnoakes commented on src/NuGet.Core/NuGet.Packaging/RuntimeModel/RuntimeGraph.cs:
```diff
@@ -8,98 +8,164 @@
-
-        private readonly ConcurrentDictionary<string, HashSet<string>> _expandCache
-            = new ConcurrentDictionary<string, HashSet<string>>(StringComparer.Ordinal);
+        private static readonly ReadOnlyDictionary<string, RuntimeDescription> EmptyRuntimes = new(new Dictionary<string, RuntimeDescription>());
+        private static readonly ReadOnlyDictionary<string, CompatibilityProfile> EmptySupports = new(new Dictionary<string, CompatibilityProfile>());
```
These singleton empty dictionaries are used in constructor chaining (to a new constructor) instead of passing `Enumerable.Empty<>` to the prior constructor which would then construct a new empty dictionary. Actually, it would do that twice, as it then wrapped it in `ReadOnlyDictionary<>`.

drewnoakes commented on src/NuGet.Core/NuGet.Packaging/RuntimeModel/RuntimeGraph.cs:
```diff
@@ -8,98 +8,164 @@
 
-        private readonly ConcurrentDictionary<RuntimeDependencyKey, List<RuntimePackageDependency>> _dependencyCache
-            = new ConcurrentDictionary<RuntimeDependencyKey, List<RuntimePackageDependency>>();
+        // These fields are null when IsEmpty is true
+        private readonly ConcurrentDictionary<RuntimeCompatKey, bool>? _areCompatible;
+        private readonly ConcurrentDictionary<string, HashSet<string>>? _expandCache;
+        private readonly ConcurrentDictionary<RuntimeDependencyKey, List<RuntimePackageDependency>>? _dependencyCache;
```
These objects are heavy, so they're no longer allocated for empty graphs.

drewnoakes commented on src/NuGet.Core/NuGet.Packaging/RuntimeModel/RuntimeGraph.cs:
```diff
@@ -8,98 +8,164 @@
+                {
+                    clone[pair.Key] = cloneFunc(pair.Value);
+                }
+
+                return clone;
+            }
         }
 
         /// <summary>
-        /// Merges the content of the other runtime graph in to this runtime graph
+        /// Merges the content of two runtime graphs and returns the combined graph.
         /// </summary>
         /// <param name="other">The other graph to merge in to this graph</param>
         public static RuntimeGraph Merge(RuntimeGraph left, RuntimeGraph right)
         {
-            var runtimes = new Dictionary<string, RuntimeDescription>();
-            foreach (var runtime in left.Runtimes.Values)
+            if (left.IsEmpty)
+            {
+                return right;
+            }
+
+            if (right.IsEmpty)
             {
-                runtimes[runtime.RuntimeIdentifier] = runtime.Clone();
+                return left;
+            }
+
+            var runtimes = new Dictionary<string, RuntimeDescription>(capacity: left.Runtimes.Count + right.Runtimes.Count);
+
+            // TODO append .NoAllocEnumerate() here once https://github.com/NuGet/NuGet.Client/pull/5246 is available
+            foreach (var pair in left.Runtimes)
+            {
+                runtimes[pair.Key] = pair.Value.Clone();
             }
 
             // Merge the right-side runtimes
-            foreach (var runtime in right.Runtimes.Values)
+            // TODO append .NoAllocEnumerate() here once https://github.com/NuGet/NuGet.Client/pull/5246 is available
+            foreach (var pair in right.Runtimes)
```
Calling `.Values` on a `Dictionary<>` allocates a `ValueCollection`. Better to just enumerate the key/value pairs together.

drewnoakes commented on src/NuGet.Core/NuGet.Packaging/RuntimeModel/RuntimeGraph.cs:
```diff
@@ -8,98 +8,164 @@
+                return left;
+            }
+
+            var runtimes = new Dictionary<string, RuntimeDescription>(capacity: left.Runtimes.Count + right.Runtimes.Count);
```
Specify the size on the dictionary to avoid resizing. In this case we know the exact size required, so use it.

drewnoakes commented on src/NuGet.Core/NuGet.Packaging/RuntimeModel/RuntimeGraph.cs:
```diff
@@ -8,98 +8,164 @@
+            var supports = new Dictionary<string, CompatibilityProfile>(capacity: right.Supports.Count + left.Supports.Count);
+
             // Copy over the right ones
-            var supports = new Dictionary<string, CompatibilityProfile>();
```
Specify the size on the dictionary to avoid resizing. In this case it's an upper bound, as `left.Supports` is filtered to only those having restore contexts. If that's uncommon, we could reduce the left's contribution to the capacity, perhaps dividing it in half.

jeffkl commented on src/NuGet.Core/NuGet.Packaging/RuntimeModel/RuntimeGraph.cs:
```diff
@@ -8,98 +8,164 @@
+
             // Copy over the right ones
-            var supports = new Dictionary<string, CompatibilityProfile>();
+            // TODO append .NoAllocEnumerate() here once https://github.com/NuGet/NuGet.Client/pull/5246 is available
```
Don't forget about this TODO

drewnoakes replied:
https://github.com/NuGet/NuGet.Client/pull/5246 has now merged. I've rebased and addressed these TODOs. cc @nkolev92

jeffkl reviewed (changes requested):
Looks good, I'll approve when https://github.com/NuGet/NuGet.Client/pull/5246 is merged and this PR is rebased with the TODO removed

nkolev92 commented:
I'll wait on the other PR for review. You can ping me when the rebase happens.

drewnoakes committed:
Use NoAllocEnumerate

kartheekp-ms commented on src/NuGet.Core/NuGet.Packaging/RuntimeModel/RuntimeGraph.cs:
```diff
@@ -197,33 +270,29 @@ public IEnumerable<RuntimePackageDependency> FindRuntimeDependencies(string runt
+                        }
                     }
                 }
+                return new List<RuntimePackageDependency>();
```
nit: Can this be `return new List<RuntimePackageDependency>(0);`?

drewnoakes replied:
Specifying zero and specifying no capacity are equivalent, although passing zero does do slightly more work as it has to validate the argument and test whether it's zero. Passing no argument is clearer and marginally cheaper.

https://github.com/microsoft/referencesource/blob/51cf7850defa8a17d815b4700b67116e3fa283c2/mscorlib/system/collections/generic/list.cs#L49-L68

davkean commented:
> Today, I learned about the concept of a readonly struct. According to the documentation, a readonly struct can undergo [boxing and unboxing conversions](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/struct#conversions) when it implements an interface. In this case, the RuntimeCompatKey is used as a key in a dictionary, and it implements the IEquatable interface. Therefore, I would anticipate that there might be some boxing operations involved when comparing keys for equality. Is it acceptable for these allocations to occur due to boxing?

No, there will be no allocations when comparing keys for equality, the call chain (through `Comparer<T>.Default`) to perform the comparison is generically specialized around 'T', not around the interface which avoids the box.

drewnoakes commented:
> Looks good, I'll approve when #5246 is merged and this PR is rebased with the TODO removed

@jeffkl the TODO has been addressed. Please take another look or dismiss your review.

drewnoakes committed:
Merge remote-tracking branch 'upstream/dev' into dev-drnoakes-fix-1826712-runtime-graph-empty-clone-merge

drewnoakes commented:
Merged in `dev` and resolved a few basic conflicts (we no longer have to clone runtimes due to #5249).

drewnoakes commented:
> > Looks good, I'll approve when #5246 is merged and this PR is rebased with the TODO removed
> 
> @jeffkl the TODO has been addressed. Please take another look or dismiss your review.

@jeffkl gentle ping
